### ISSUE TITLE
Pyramid 1.8.x and 1.9.x installs release candidate version of WebOb

### ISSUE THREAD

**Issue by hyperknot** [Pylons/pyramid#3220]
Since WebOb 1.8.0rc1 is released, Pyramid 1.8 and 1.9 versions install a the rc version of WebOb, instead of the latest stable 1.7.4.

**1.7**
```
pip install 'pyramid < 1.8'
Collecting WebOb>=1.3.1 (from pyramid<1.8)
  Using cached WebOb-1.7.4-py2.py3-none-any.whl
```

**1.8**
```
pip install 'pyramid < 1.9'
Collecting WebOb>=1.7.0rc2 (from pyramid<1.9)
  Using cached WebOb-1.8.0rc1-py2.py3-none-any.whl
```

**1.9**
```
pip install 'pyramid' 
Collecting WebOb>=1.7.0rc2 (from pyramid)
  Using cached WebOb-1.8.0rc1-py2.py3-none-any.whl
```

This behaviour is exactly the same with pip-tools / pip-compile.

A simple `pip install webob` installs the stable 1.7.4.

Is this a bug in pip?

**Comment by bertjwregeer** [comment]
What version of pip are you using? It shouldn't be installing the rc version of 1.8.0.

I am seeing the same issue though with the latest version of pip:

```(testvenv) Riley:~ xistence$ pip install pyramid
Collecting pyramid
  Using cached pyramid-1.9.1-py2.py3-none-any.whl
Collecting WebOb>=1.7.0rc2 (from pyramid)
  Downloading WebOb-1.8.0rc1-py2.py3-none-any.whl (114kB)
    100% |████████████████████████████████| 122kB 1.3MB/s 
Collecting repoze.lru>=0.4 (from pyramid)
  Using cached repoze.lru-0.7-py3-none-any.whl
Requirement already satisfied: setuptools in ./testvenv/lib/python3.6/site-packages (from pyramid)
Collecting plaster-pastedeploy (from pyramid)
  Using cached plaster_pastedeploy-0.4.2-py2.py3-none-any.whl
Collecting plaster (from pyramid)
  Using cached plaster-1.0-py2.py3-none-any.whl
Collecting zope.deprecation>=3.5.0 (from pyramid)
  Using cached zope.deprecation-4.3.0-py2.py3-none-any.whl
Collecting zope.interface>=3.8.0 (from pyramid)
Collecting venusian>=1.0a3 (from pyramid)
  Using cached venusian-1.1.0-py2.py3-none-any.whl
Collecting PasteDeploy>=1.5.0 (from pyramid)
  Using cached PasteDeploy-1.5.2-py2.py3-none-any.whl
Collecting hupper (from pyramid)
  Using cached hupper-1.0-py2.py3-none-any.whl
Collecting translationstring>=0.4 (from pyramid)
  Using cached translationstring-1.3-py2.py3-none-any.whl
Installing collected packages: WebOb, repoze.lru, plaster, PasteDeploy, plaster-pastedeploy, zope.deprecation, zope.interface, venusian, hupper, translationstring, pyramid
Successfully installed PasteDeploy-1.5.2 WebOb-1.8.0rc1 hupper-1.0 plaster-1.0 plaster-pastedeploy-0.4.2 pyramid-1.9.1 repoze.lru-0.7 translationstring-1.3 venusian-1.1.0 zope.deprecation-4.3.0 zope.interface-4.4.3```

**Comment by bertjwregeer** [comment]
Side note: Pyramid 1.9 will work just fine with the latest WebOb RC, and I'd appreciate the testing ;-)

**Comment by hyperknot** [comment]
pip 9.0.1

I really think it has to be a bug with pip. Should Pyramid 1.8 also work fine? Anyway I prefer staying on stable versions.

**Comment by hyperknot** [comment]
Actually the minimal case is as simple as: 
```
pip install 'webob >= 1.7.0rc2'
Collecting webob>=1.7.0rc2
  Using cached WebOb-1.8.0rc1-py2.py3-none-any.whl
```

**Comment by bertjwregeer** [comment]
Pyramid 1.8 will also work fine with WebOb 1.8rc1, all of the tests pass. Pyramid 1.9 is the latest stable Pyramid release.

**Comment by hyperknot** [comment]
Opened pip issue: https://github.com/pypa/pip/issues/4969

**Comment by mmerickel** [comment]
I'm fairly certain this is "expected" behavior and it's pip relaxing its constraints when upgrading a non-stable version. If you were on 1.7.0 instead of 1.7.0rc2 it would work as expected. Anyway this isn't a pyramid issue so I'm closing it but I look forward to seeing what the pip folks have to say.

**Comment by mmerickel** [comment]
If the issue is that pyramid itself is pinned to a non-stable version then I certainly think that's a pip issue that should be fixed. I originally thought it was because you had the non-stable webob installed already but I think I read that wrong.

**Comment by hyperknot** [comment]
@mmerickel just a small note, I was always running it in a clean virtualenv, without upgrading anything.

**Comment by hyperknot** [comment]
pip is working as intended, which is different from how npm handles semver ranges.
See details in the pip issue: https://github.com/pypa/pip/issues/4969

If Pyramid wants to avoid having 1.8 and 1.9 users try out every single pre-release version in the future (including 1.9-alpha, 2.0-rc1, etc.), the dependencies have to be modified for a stable release, for example 1.7.0.

Until that, users need to pin webob explicitly to < 1.8 for example.

**Comment by mmerickel** [comment]
Yeah this is quite surprising behavior to me. I have re-opened this ticket and since pip is broken (imo) we will need to release newer versions of 1.8/1.9 that depend on non-pre-release versions of webob and any other dependencies because of course Pyramid never intended that its users should always use pre-release versions of webob even though pip seems to think that this specifier is equivalent to `--pre`.

**Comment by mmerickel** [comment]
I'm loathe to release bugfix versions of Pyramid that change dependency requirements. My current plan is to fix this in the upcoming 1.10 release instead.

**Comment by bertjwregeer** [comment]
This was fixed in Pyramid 1.9.2 that was released late last night.
